example_tcp_client/src/main.rs

```diff
@@ -100,8 +128,44 @@ fn write_to_kanata(mut s: TcpStream) {
     loop {
         stdin().read_line(&mut layer).expect("stdin is readable");
         let new = layer.trim_end().to_owned();
-        if new.starts_with("fk:") {
-            let fkname = new.trim_start_matches("fk:").into();
+        if new.starts_with("fkpress:") {
+            let fkname = new.trim_start_matches("fkpress:").into();
+            log::info!("writer: telling kanata to press fake key \"{fkname}\"");
+            let msg = serde_json::to_string(&ClientMessage::ActOnFakeKey {
+                name: fkname,
+                action: FakeKeyActionMessage::Press,
+            })
+            .expect("deserializable");
+            s.write_all(msg.as_bytes()).expect("stream writable");
+            layer.clear();
+            continue;
+        }
+        if new.starts_with("fkrelease:") {
+            let fkname = new.trim_start_matches("fkrelease:").into();
+            log::info!("writer: telling kanata to release fake key \"{fkname}\"");
+            let msg = serde_json::to_string(&ClientMessage::ActOnFakeKey {
+                name: fkname,
+                action: FakeKeyActionMessage::Release,
+            })
+            .expect("deserializable");
+            s.write_all(msg.as_bytes()).expect("stream writable");
+            layer.clear();
+            continue;
+        }
+        if new.starts_with("fktap:") {
```

Comment:
Did you intend this to be toggle? There are two taps

Reply:
I don't think I intended to commit that change (sorry, in the process of learning git). I've fixed it on my end and can commit+push, but I felt like there should be a cleaner way to handle the fk___: entries without so much repeated code. Seems like that could be abstracted into a `handle_fake_keys` function along the lines of the below (which doesn't compile because wow indexing strings is way harder than in C).
```rust
fn lookup_fakekeyactionmessage(fake_key_action_name: &str) -> FakeKeyActionMessage {
    // dummy function; could be my old TryFromStr code
}

fn handle_fake_key_cmd(new: &str) {
    // if it were Python I'd do a regex match here to look for r"fk\w+:", not sure about Rust
    // Could always do if ...starts_with("fkpress:") || ...starts_with("fkrelease:") || ...
    let colon_index = new.find(":").expect("contains ':'");
    let action = new.skip(colon_index).collect(); // Best practice to assert colon_index >= 2 here?
    let fkname = new[colon_index..];
    log::info!("writer: telling kanata to {action} fake key \"{fkname}\"");
    let msg = serde_json::to_string(&ClientMessage::ActOnFakeKey {
        name: fkname,
        action: lookup_fakekeyactionmessage(action),
    })
    .expect("deserializable");
    s.write_all(msg.as_bytes()).expect("stream writable");
    layer.clear();
}
```

Reply:
Here's how I would do it:

https://github.com/jtroo/kanata/commit/7c414aca9e684252210f4b07a6b1b06da16fb8d7

Reply:
Sum types (e.g. `Option`), `match`, and standard library convenience functions are your friend :)

Reply:
Thank you for the example! I haven't forgotten about this, just a busy few days and I'm trying to get some practice before circling back to this. I should have some new code before too long. That `split_once` function sure simplifies things from what I was doing in my scratch code above.